components/security-mgt/org.wso2.carbon.security.mgt/src/main/java/org/wso2/carbon/security/keystore/KeyStoreManagementServiceImpl.java

```diff
@@ -180,12 +217,262 @@ public void addCertificate(String tenantDomain, String alias, String certificate
     public void deleteCertificate(String tenantDomain, String alias) throws KeyStoreManagementException {
 
         try {
-            getKeyStoreAdmin(tenantDomain).removeCertFromStore(alias, getKeyStoreName(tenantDomain));
+            KeyStoreAdmin keyStoreAdmin = getKeyStoreAdmin(tenantDomain);
+            String keystoreName = getKeyStoreName(tenantDomain);
+            if (!isPrivateKey(keyStoreAdmin, keystoreName, alias)) {
+                keyStoreAdmin.removeCertFromStore(alias, keystoreName);
+            }
+        } catch (Exception e) {
+            throw handleServerException(ERROR_CODE_DELETE_CERTIFICATE, alias, e);
+        }
+    }
+
+    public void addPrivateKey(String alias, String privateKeyContent, String certificateChain, String tenantDomain) throws KeyStoreManagementException {
+
+        if (tenantDomain.equals(MultitenantConstants.SUPER_TENANT_DOMAIN_NAME)) {
+            log.error("Adding private key for super tenant is not supported");
+            throw handleClientException(ERROR_CODE_UNSUPPORTED_ADDING_KEYS_FOR_SUPER_TENANT, alias);
+        }
+        String keyStoreName = getKeyStoreName(tenantDomain);
+        KeyStore keyStore;
+        KeyStoreManager keyStoreManager;
+        try {
+            keyStoreManager = getKeyStoreManager(tenantDomain);
+            keyStore = keyStoreManager.getKeyStore(keyStoreName);
+        } catch (Exception e) {
+            throw handleServerException(ERROR_CODE_ADD_PRIVATE_KEY, alias, e);
+        }
+        boolean isKeyStoreContainsAlias = false;
+        try {
+            isKeyStoreContainsAlias = keyStore.containsAlias(alias);
+        } catch (Exception e) {
+            throw handleServerException(ERROR_CODE_ADD_PRIVATE_KEY, alias, e);
+        }
+        if (isKeyStoreContainsAlias) {
+            if (log.isDebugEnabled()) {
+                log.debug("Keystore of the tenant: " + tenantDomain + "already contains an entry with the " +
+                        "alias:" + " " + alias);
+            }
+            throw handleClientException(ERROR_CODE_AN_ENTRY_FOR_THE_GIVEN_ALIAS_EXISTS, alias);
+        }
+        try {
+            String password = keyStoreManager.getKeyStorePassword(keyStoreName);
+            Key privateKey = KeyStoreMgtUtil.extractPrivateKey(privateKeyContent);
+            Certificate[] x509CertificateChain = extractCertificateChain(certificateChain);
+            keyStore.setKeyEntry(alias, privateKey, password.toCharArray(), x509CertificateChain);
+            updateKeyStore(keyStore, tenantDomain, password);
+        } catch (Exception e) {
```

Comment:
Can we catch specific exception, not the generic Exception class here ?

Reply:
keyStoreManager.getKeyStorePassword(keyStoreName) is throwing a generic Exception.